Kernel/TTY/VirtualConsole.cpp

```diff
@@ -1,129 +1,58 @@
 /*
  * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
  * Copyright (c) 2020, Sergey Bugaev <bugaevc@serenityos.org>
+ * Copyright (c) 2021, Liav A. <liavalb@hotmail.co.il>
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #include "VirtualConsole.h"
+#include <AK/StdLibExtras.h>
 #include <AK/String.h>
 #include <Kernel/Arch/x86/CPU.h>
+#include <Kernel/Debug.h>
 #include <Kernel/Devices/HID/HIDManagement.h>
+#include <Kernel/Graphics/GraphicsManagement.h>
 #include <Kernel/Heap/kmalloc.h>
 #include <Kernel/IO.h>
 #include <Kernel/StdLib.h>
+#include <Kernel/TTY/ConsoleManagement.h>
 
 namespace Kernel {
 
-static u8* s_vga_buffer;
-static VirtualConsole* s_consoles[s_max_virtual_consoles];
-static int s_active_console;
-static RecursiveSpinLock s_lock;
-
-void VirtualConsole::flush_vga_cursor()
-{
-    u16 value = m_current_vga_start_address + (m_terminal.cursor_row() * columns() + m_terminal.cursor_column());
-    IO::out8(0x3d4, 0x0e);
-    IO::out8(0x3d5, MSB(value));
-    IO::out8(0x3d4, 0x0f);
-    IO::out8(0x3d5, LSB(value));
-}
-
-UNMAP_AFTER_INIT void VirtualConsole::initialize()
+void VirtualConsole::set_graphical(bool graphical)
 {
-    s_vga_buffer = (u8*)0xc00b8000;
-    s_active_console = -1;
+    m_graphical = graphical;
 }
 
-void VirtualConsole::set_graphical(bool graphical)
+UNMAP_AFTER_INIT NonnullRefPtr<VirtualConsole> VirtualConsole::create(size_t index)
 {
-    if (graphical)
-        set_vga_start_row(0);
-
-    m_graphical = graphical;
+    return adopt_ref(*new VirtualConsole(index));
 }
 
 UNMAP_AFTER_INIT VirtualConsole::VirtualConsole(const unsigned index)
     : TTY(4, index)
     , m_index(index)
-    , m_terminal(*this)
 {
-    VERIFY(index < s_max_virtual_consoles);
-
     m_tty_name = String::formatted("/dev/tty{}", m_index);
-    m_terminal.set_size(80, 25);
+    VERIFY(GraphicsManagement::the().console());
+    change_size(GraphicsManagement::the().console()->max_column(), GraphicsManagement::the().console()->max_row());
+    // 2 mebibytes seems like a reasonable amount of Cells to store within...
+    m_cells = MM.allocate_kernel_region(2 * MiB, "Virtual Console Cells", Region::Access::Read | Region::Access::Write, AllocationStrategy::AllocateNow);
```

Comment:
Couldn't this be calculated?

Reply:
Probably yes. Let me look into a solution :)